File: ghrah-observer-webui/packages/protocol/src/payloads/manifest.ts
import { z } from "zod";

// ─── Manifest CRUD Payloads ───

export const ManifestListPayloadSchema = z.object({
  namespace: z.string().nullable().optional(),
});

export const ManifestGetPayloadSchema = z.object({
  full_name: z.string(),
});

export const ManifestPutPayloadSchema = z.object({
  full_name: z.string(),
  content: z.string(),
  overwrite: z.boolean().optional().default(false),
});

export const ManifestDeletePayloadSchema = z.object({
  full_name: z.string(),
});

export const ManifestValidatePayloadSchema = z.object({
  content: z.string(),
  manifest_type: z.string(),
});

export const ManifestResolvePayloadSchema = z.object({
  agent_full_name: z.string(),
  runtime_name: z.string().nullable().optional(),
});

// ─── Manifest Event Payloads ───

export const ManifestAbilityEventPayloadSchema = z.object({
  full_name: z.string(),
  namespace: z.string(),
  manifest: z.record(z.unknown()).nullable().optional(),
  source: z.string().nullable().optional(),
});

export const ManifestAgentEventPayloadSchema = z.object({
  full_name: z.string(),
  namespace: z.string(),
  manifest: z.record(z.unknown()).nullable().optional(),
  source: z.string().nullable().optional(),
});

export type ManifestListPayload = z.infer<typeof ManifestListPayloadSchema>;
export type ManifestGetPayload = z.infer<typeof ManifestGetPayloadSchema>;
export type ManifestPutPayload = z.infer<typeof ManifestPutPayloadSchema>;
export type ManifestDeletePayload = z.infer<typeof ManifestDeletePayloadSchema>;
export type ManifestValidatePayload = z.infer<typeof ManifestValidatePayloadSchema>;
export type ManifestResolvePayload = z.infer<typeof ManifestResolvePayloadSchema>;
export type ManifestAbilityEventPayload = z.infer<typeof ManifestAbilityEventPayloadSchema>;
export type ManifestAgentEventPayload = z.infer<typeof ManifestAgentEventPayloadSchema>;

File: ghrah-protocol/src/ghrah/protocol/payloads/manifest.py
# SPDX-FileCopyrightText: 2026 chenxya <chenxya@ghrah.org>
#
# SPDX-License-Identifier: Apache-2.0
"""Manifest 域载荷模型（CRUD 命令、响应与事件）。"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel

# ─── Manifest CRUD 载荷模型 ───


class ManifestListPayload(BaseModel):
    """manifest_list_abilities / manifest_list_agents 命令载荷。"""

    namespace: str | None = None


class ManifestGetPayload(BaseModel):
    """manifest_get_ability / manifest_get_agent 命令载荷。"""

    full_name: str


class ManifestPutPayload(BaseModel):
    """manifest_put_ability / manifest_put_agent 命令载荷。"""

    full_name: str
    content: str
    overwrite: bool = False


class ManifestDeletePayload(BaseModel):
    """manifest_delete_ability / manifest_delete_agent 命令载荷。"""

    full_name: str


class ManifestValidatePayload(BaseModel):
    """manifest_validate 命令载荷。"""

    content: str
    manifest_type: str


class ManifestResolvePayload(BaseModel):
    """manifest_resolve_agent 命令载荷。"""

    agent_full_name: str
    runtime_name: str | None = None  # ─── Manifest CRUD 响应载荷模型 ───


class ManifestGetAbilityResponsePayload(BaseModel):
    """manifest_get_ability 响应载荷。"""

    manifest: dict[str, Any]
    source: str


class ManifestGetAgentResponsePayload(BaseModel):
    """manifest_get_agent 响应载荷。"""

    manifest: dict[str, Any]
    source: str


class ManifestPutResponsePayload(BaseModel):
    """manifest_put_ability / manifest_put_agent 响应载荷。"""

    full_name: str


class ManifestDeleteResponsePayload(BaseModel):
    """manifest_delete_ability / manifest_delete_agent 响应载荷。"""

    full_name: str


class ManifestValidateResponsePayload(BaseModel):
    """manifest_validate 响应载荷。"""

    valid: bool
    errors: list[str]


class ManifestResolveAgentResponsePayload(BaseModel):
    """manifest_resolve_agent 响应载荷。"""

    config: dict[str, Any]
    abilities: list[dict[str, Any]]  # ─── Manifest 事件载荷模型 ───


class ManifestAbilityEventPayload(BaseModel):
    """manifest_ability_created/updated/deleted 事件载荷。"""

    full_name: str
    namespace: str
    manifest: dict[str, Any] | None = None
    source: str | None = None


class ManifestAgentEventPayload(BaseModel):
    """manifest_agent_created/updated/deleted 事件载荷。"""

    full_name: str
    namespace: str
    manifest: dict[str, Any] | None = None
    source: str | None = None
